### PR TITLE
Signing: assert package signature file is a regular file

### LINKED ISSUE #6343: Signing:  assert package signature file is a regular file
PR https://github.com/NuGet/NuGet.Client/pull/1871 added the package signature file to a package.  This change asserts that the package signature file is a regular file and not some other file type (e.g.:  directory, symbolic link, etc.) that ZIP's external file attributes permit.

### PR DESCRIPTION
**Pull request by dtivel**
Fix https://github.com/NuGet/Home/issues/6343.

Also:

* fix incorrect decoding of ZIP entry file names
* simplify ZIP-reading code

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/Signing/Archive/CentralDirectoryHeader.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        internal ushort FileCommentLength { get; private set; }
+        internal ushort DiskNumberStart { get; private set; }
+        internal ushort InternalFileAttributes { get; private set; }
+        internal uint ExternalFileAttributes { get; private set; }
+        internal uint RelativeOffsetOfLocalHeader { get; private set; }
+        internal byte[] FileName { get; private set; }
+        internal byte[] ExtraField { get; private set; }
+        internal byte[] FileComment { get; private set; }
+
+        internal uint GetSize()
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Archive/CentralDirectoryHeader.cs]: given that this size doesn't include signature, should we change the name or add a comment to specify that is size of the header without signature?

**Review by PatoBeltran — APPROVED** [review]
Nice job! Just one minor comment. Make sure to fix failing functional tests before merging

File: src/NuGet.Core/NuGet.Packaging/Signing/Archive/CentralDirectoryHeader.cs (hunk 1 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/Signing/Archive/CentralDirectoryHeader.cs]: Actually, `SignedPackageArchiveIOUtility.CentralDirectoryFileHeaderSizeWithoutSignature` is misnamed.  Its value (46) includes the signature.  I'll fix this though.

```
Field                             Bytes
--------------------------------- ------
uint Signature                    4
ushort VersionMadeBy              2
ushort VersionNeededToExtract     2
ushort GeneralPurposeBitFlag      2
ushort CompressionMethod          2
ushort LastModFileTime            2
ushort LastModFileDate            2
uint Crc32                        4
uint CompressedSize               4
uint UncompressedSize             4
ushort FileNameLength             2
ushort ExtraFieldLength           2
ushort FileCommentLength          2
ushort DiskNumberStart            2
ushort InternalFileAttributes     2
uint ExternalFileAttributes       4
uint RelativeOffsetOfLocalHeader  4
--------------------------------- ------
Total                             46
```

**Review by emgarten — APPROVED** [review]
LGTM 🚀

**Review by jainaashish — APPROVED** [review]

**Pull request merged**